lib/classes/Service.js

```diff
@@ -97,10 +97,11 @@ class Service {
 
     if (_.isObject(serverlessFile.service)) {
       that.serviceObject = serverlessFile.service;
-      that.service = serverlessFile.service.name;
+      if (options.service) that.serviceObject.name = options.service;
+      that.service = options.service || serverlessFile.service.name;
```

Comment:
This line can stay as it was (as `serverlessFile.service` is `that.serviceObject`).

Probably least confusing way to put it this whole condition would be:
```javascript
if (_.isObject(serverlessFile.service)) this.serviceObject = serverlessFile.service;
else that.serviceObject = { name: options.service || serverlessFile.service };

that.service =  = that.serviceObject.name;
```

Reply:
> What exactly would be "targeted instruction" in the dashboard?

Targeted instructions refers to the instructions [here](https://serverless.invisionapp.com/share/FETSG5YVM4R#/382360490_app-Service-List-Empty) and [here](https://serverless.invisionapp.com/share/FETSG5YVM4R#/382360491_service-Empty). As you can see, the intent is to enable the user to use the `serverless` wizard to create a new project and associate it with a specific organization, application and service which already exists in the Dashboard but does not exist on the CLI yet.

Reply:
If you run the `serverless` command on an existing project and the value provided with the `--service` flag doesn't match the value of the `serverless.yml` I'd suggest updating the value in the `serverless.yml`, but showing a warning, `The service name in the serverless.yml has been updated from "${pre}" to "${post}"`

Reply:
> the intent is to enable the user to use the serverless wizard to create a new project and associate it with a specific organization, application 

Ok, but isn't that supposed to not be mixed with interactive CLI experience? e.g. if one runs:

```
sls --org some-org --app some-app
```

I would expect that service gets automatically updated to provided settings, and no interactive questions are asked.

Reply:
Elaborating on all cases:

(1) When user wants to update either _org_, _app_ and/or _service name_ in existing service, there's nothing easier than updating it in `serverless.yml`. I doubt developers would prefer to propagate such basic updates via CLI command.

(2) When user wants to create a new service with predefined settings, we have `sls create` that's dedicated for that, and it seems to me, that's where the update should be made.
Once we add there support for `org` and `app`, service setup command may look as

```
sls create -t aws-nodejs --name my-new-service --org my-org --app my-app-01
```
We may further improve it, by making`-t` optional, and by default assume `aws-nodejs` (as it's most common scenario). Then service setup command will look as:

```
sls create -name my-new-service --org my-org --app my-app-01
```

(3) When user wants to _interactively_ create/update the service, already existing means seem to address intent very well. Why user when choosing an _interactive_ setup, may want to pass value via CLI, instead of choosing it from a presented list (usually just by clicking enter)?

Also in case of interactive CLI, there are configuration scenarios for which `--org` and `--app` will be invalid (not really applicable). To make it solid either we will have to report that such options cannot be applied, or filter out scenarios which cannot apply, and that will make the implementation even more complex (which isn't free), while value added remains questionable (imo)

Reply:
(1) As a user, if you already have a service, you are already registered, you are already logged in via CLI, you are already familiar with how the dashboard is integrated, and you have your AWS credentials setup, then you would probably just update the `serverless.yml` in your IDE instead of using the `serverless` CLI. But if any one of those conditions is not true, then it's probably easier to run `serverless`.

(2) The `sls create` command does not handle the AWS credential setup. That's a big hurdle we wish to avoid.

(3) It is easier to instruct the user to run a single command. In fact, they can just hit the copy button in the dashboard and paste it in the CLI, and it'll work as expected. Instructing them to select the organization, application, and then later updating the `serverless.yml` adds a lot of error prone steps. 

> Also in case of interactive CLI, there are configuration scenarios for which --org and --app will be invalid (not really applicable).
In many of those cases they can just be ignored. Correct me if I am wrong, but those flags won't be documented when you run `sls help`. So the only time they will discover those flags will be when they are instructed to use them in the dashboard.

Reply:
Ok, I think I need to better understand linked mocks, which expose targeted instructions. They're not perfectly to me

1. https://serverless.invisionapp.com/share/FETSG5YVM4R#/screens/382360490_app-Service-List-Empty

This I take is services page, at stage when none of the services are registered for given app.

It lists the steps on how to pre-configure new (or update existing) serverless service.

And `--service` option (not sure why not listed here, it could be as `--service <new-service-name>`), is expected to be used when new service is to be created (at least that's the use case for which it's proposed)

I take we decided to not support an extra options, through which AWS creds setup will be preconfigured, as it's not service specific. Still we strongly want the user to also be guided through AWS creds setup (assuming they're not present in his environment) hence we expect the interactive setup to jump in.

Is that correct?

2. https://serverless.invisionapp.com/share/FETSG5YVM4R#/screens/382360491_service-Empty

This one is not perfectly clear to me. Do we plan to allow to "create" service in dashboard context, which technically will just mean, that service is scheduled for creation (as that wouldn't initialize any project repository) ?

If it's the case, then I believe mock shows the page of just _registered_ service, but which was not yet really created (or we don't know it was, as no deployments were registered) (?)

Having that, targetted instructions are not that different from (1), aside of fact that we're rather addressing case on how to _create_ a new project (not update existing), and prefill `--service` with specific value.

I this correct?

_One important note: In `sls create` we already have an option through which we propose service name, and it's `--name`. To maintain CLI consistency, I believe we should also  use `--name` here (and not `--service`)_

Reply:
> One important note: In sls create we already have an option through which we propose service name, and it's --name. To maintain CLI consistency, I believe we should also use --name here (and not --service)

Good point!

Reply:
> The service name in the serverless.yml has been updated from "${pre}" to "${post}"

is this on interactive calls or somethign like deploy?

Reply:
> And --service option (not sure why not listed here, it could be as --service <new-service-name>), is expected to be used when new service is to be created (at least that's the use case for which it's proposed)

It isn't listed because that page is in the context of the application, not service, so there is no service at the time. That is, the user has not yet created a service in the dashboard. By not specifying the `name` (what used to be `service`) flag, they will be prompted for it in the interactive flow, which is the intended behavior.

> I take we decided to not support an extra options, through which AWS creds setup will be preconfigured, as it's not service specific. Still we strongly want the user to also be guided through AWS creds setup (assuming they're not present in his environment), and this is the sole reason, for which we want the interactive setup to jump in.

If I understand the question, you are asking to confirm that we are not adding any new flags for the AWS credentials. Yes, that is correct, this does not introduce any new flags for the AWS credentials. The intent of the flags is not to replace the need for interactivity, it is to set the context to align it to the org/app/service with the dashboard. 

> Do we plan to allow to "create" service in dashboard context, which technically will just mean, that service is scheduled for creation (as that wouldn't initialize any project repository) ?

Yes. Currently services can only be created via a deployment. This adds support for creating services in the dashboard.

> Having that, targetted instructions are not that different from (1), aside of fact that we're rather addressing case on how to create a new project (not update existing), and prefill --service with specific value.

You got it. The two instructions are very similar; however, they appear in different contexts. The first appears in the context of the application, not the service, and does not specify the `--name` flag, the other one is in the context of a service with no service instances and specifies the `--name` flag.

> One important note: In sls create we already have an option through which we propose service name, and it's --name. To maintain CLI consistency, I believe we should also use --name here (and not --service)

👍 on `name` instead of `service`.

> is this on interactive calls or something like deploy?

Interactive calls only. The flags I proposed only apply to the `sls` command, no other commands.

Reply:
> Interactive calls only. The flags I proposed only apply to the sls command, no other commands. 

ohhhhh. I thought this over rode those settings on any command.

Reply:
Thanks for clarifications @skierkowski 

> The intent of the flags is not to replace the need for interactivity, it is to set the context to align it to the org/app/service with the dashboard.

Problem of this proposal is that given `--org`, `--app` and `--name` options will have to be supported/handled in all possible interactive CLI contexts, and for interactive CLI possible contexts are quite different - in most cases they do not align with contexts presented on dashboard pages we're addressing now.

Second problem is that interactive CLI by its nature is _optionless_. Whole idea of it is that we provide values for options _interactively_. That characteristics also allows to clearly distinct it from other commands (it's how we detect that interactive CLI should jump in).

Note that there are just options (e.g. `sls --help`, `sls --verbose`) which trigger specific output.
So when introducing an options for an interactive CLI, we're starting to use same namespace for two different commands, which doesn't feel right, raises risk of conflicts and implementation issues.

When adding support for options we should also provide some help to list and describe those possible options. It's not clear how to do that, when `sls --help` already does something very different.

What I propose is that we fine tune those targeted instructions, so they align with how developers are used to work with Serverless projects, and take into account current characteristics of a CLI and its possible contexts.

What do you think of following:

_Install the Serverless Framework (if you haven't already)_
```bash
$ npm install -g serverless
```
_Login to your Serverless Framework account_
```bash
$ serverless login
```
_To integrate dashboard into an existing service update `serverless.yml` as follows:_
```yaml
org: my-org
app: my-app-01
```
_To start a new integrated service, run:_
```
$ serverless create --org my-org --app my-app-01
$ cd my-app-01
```
_In context of an existing service, you may configure other various features (e.g. preconfigure AWS services access or setup CLI autocompletion), simply by running our interactive setup:_
```
$ serverless
```
_Deploy your service:_
```bash
$ serverless deploy
```

Ensuring support for above would be quite straightforward on our side.